fix(ui): issues controlling cursor position with ext_cmdline

luukvbaal opened the pull request:
**fix(api): update grid cursor in nvim_win_set_cursor()**

Problem:  Cursor position set by nvim_win_set_cursor() is not reflected
          on the screen when followed by a blocking call like getchar().
Solution: Immediately update the cursor position on the grid.

Fix #20793

**fix(ui): ext_cmdline should not move cursor to curwin**

Problem:  The ext_cmdline cursor position on the screen seems to rely on
          an implicit assumption that the event listener implements a
          cmdline window that is made the current window which is
          problematic (e.g. breaks 'incsearch' in the actual current
          window).
Solution: Remove this assumption and allow nvim_win_set_cursor() to move
          the cursor on the screen to a non-current window (previous
          commit).

Fix #20309

bfredl commented on src/nvim/api/window.c:
```diff
@@ -134,25 +153,32 @@ void nvim_win_set_cursor(Window window, ArrayOf(Integer, 2) pos, Error *err)
+    restore_win(&switchwin, true);
+    redraw_later(win, UPD_VALID);
+    win->w_redr_status = true;
+  } else {  // screen
+    ui_grid_cursor_goto(default_grid.handle, (int)row - 1, (int)col);
+    skip_cursor_update = true;
```
just my immediate reaction, but this part seems to be _very_ different functionality, to the extent it should be a separate API call altogether.

luukvbaal replied:
I see. Like I said I didn't even have a use case for it.
I just went "no opts array" -> nvim_set_cursor() -> add 2 lines for screen relative cursor.
If it isn't wanted perhaps this should just be nvim_set_cursor2()?

zeertzjq commented:
Use (1,0) indexing for screen position is strange, as it only makes sense for text position.

luukvbaal commented:
> Using (1,0) indexing for screen position is strange.

I thought so too but using different indexing for different options in the same functions was even stranger to me. Might be irrelevent if we're not adding `screen`.

luukvbaal commented:
> Might be irrelevent if we're not adding screen.

In that case we might as well forego this new API and just resolve this by having `nvim_win_set_cursor()` call `setcursor_mayforce()`. I don't think there is any harm in that?

luukvbaal committed:
fix(api): update grid cursor in nvim_win_set_cursor()

Problem:  Cursor position set by nvim_win_set_cursor() is not reflected
          on the screen when followed by a blocking call like getchar().
Solution: Immediately update the cursor position on the grid.

luukvbaal committed:
fix(ui): ext_cmdline should not move cursor to curwin

Problem:  The ext_cmdline cursor position on the screen seems to rely on
          an implicit assumption that the event listener implements a
          cmdline window that is made the current window which is
          problematic (e.g. breaks 'incsearch' in the actual current
          window).
Solution: Remove this assumption and allow nvim_win_set_cursor() to move
          the cursor on the screen to a non-current window (previous
          commit).

luukvbaal commented on src/nvim/ex_getln.c:
```diff
@@ -767,7 +767,7 @@ static uint8_t *command_line_enter(int firstc, int count, int indent, bool clear
 
   setmouse();
-  ui_cursor_shape();               // may show different cursor shape
+  setcursor();
```
Instead, move the cursor back to curwin when leaving the cmdline (and remove duplicate call to `ui_cursor_shape()`).

zeertzjq replied:
Is removing duplicate call to `ui_cursor_shape()` really necessary? It doesn't do anything if editor state hasn't changed.

luukvbaal replied:
No but why would we keep it? It's called in setmouse() in the line above.

zeertzjq replied:
Oh, I thought you were talking about `command_line_check()`.

bfredl reviewed (approved):
Yes, I think the behavior in the added tests make a lot of sense.

echasnovski commented:
I am not sure forcing cursor draw in not-current window is good behavior. This might lead to visually indicating that current window has changed when in reality it has not. And if it is not a huge deal in regular Nvim TUI, in Neovide it might lead to [this kind of behavior](https://github.com/neovide/neovide/issues/2466#issuecomment-2041117466).

The use case here is that user input is taken one by one via `getchar()` (technically, `getcharstr()`), while calling `nvim_win_set_cursor()` on the not current window which is never meant to be focused. The `nvim_win_set_cursor()` is usually needed to ensure the proper window view: either first line as top line ('mini.clue') or cursorline highlighting (which *I think* is what is done in 'nvim-cmp').
